File: cookbooks/fraud-graph-traversal/03-fan-detection.py
"""Step 03 — Fan-in (mule) and fan-out (splitter) detection.

Pulls (src, dst) pairs from one MATCH and tallies distinct counterparts
per account in Python. The query side is trivial; the analysis side is
where business rules live (thresholds, time windows, account categories).

Why this shape — `count(DISTINCT x)` is not yet stable in WorldCypher 1.6.6,
so we keep the engine query simple and tally distincts client-side.
A future engine release flips this single step to a one-line aggregate;
the surrounding code stays as is.

Run:
    uv run python 03-fan-detection.py
"""
from __future__ import annotations

from collections import defaultdict

from _load import load

FAN_IN_THRESHOLD = 8
FAN_OUT_THRESHOLD = 8


def main() -> None:
    db, _ = load()

    result = db.execute(
        "MATCH (src:Account)-[:SENT]->(t:Transfer)-[:TO]->(dst:Account) "
        "RETURN src.acct_id AS src, dst.acct_id AS dst, t.amount AS amount"
    )
    pairs = [(row["src"], row["dst"], float(row["amount"])) for row in result]

    fan_in: dict[str, set[str]] = defaultdict(set)
    fan_out: dict[str, set[str]] = defaultdict(set)
    volume_in: dict[str, float] = defaultdict(float)
    volume_out: dict[str, float] = defaultdict(float)
    for src, dst, amount in pairs:
        fan_in[dst].add(src)
        fan_out[src].add(dst)
        volume_in[dst] += amount
        volume_out[src] += amount

    print("\n[1] Fan-in concentration — accounts receiving from many sources:")
    top_in = sorted(fan_in.items(), key=lambda kv: -len(kv[1]))[:5]
    for acct, sources in top_in:
        flag = "  <-- FLAG" if len(sources) >= FAN_IN_THRESHOLD else ""
        print(
            f"    {acct:>12}  distinct_sources={len(sources):>3}  "
            f"total=${volume_in[acct]:>10,.2f}{flag}"
        )

    print("\n[2] Fan-out layering — accounts sending to many destinations:")
    top_out = sorted(fan_out.items(), key=lambda kv: -len(kv[1]))[:5]
    for acct, dsts in top_out:
        flag = "  <-- FLAG" if len(dsts) >= FAN_OUT_THRESHOLD else ""
        print(
            f"    {acct:>12}  distinct_destinations={len(dsts):>3}  "
            f"total=${volume_out[acct]:>10,.2f}{flag}"
        )

    print(f"\n[3] Drill-down on planted patterns:")
    print(
        f"    MULE-1     received from {len(fan_in['MULE-1'])} distinct sources, "
        f"total=${volume_in['MULE-1']:,.2f}"
    )
    print(
        f"    SPLITTER-1 sent to        {len(fan_out['SPLITTER-1'])} distinct dsts,    "
        f"total=${volume_out['SPLITTER-1']:,.2f}"
    )

    db.close()
    print("\nOK")


if __name__ == "__main__":
    main()

File: examples/fraud-graph-traversal/rust/src/bin/03-fan-detection.rs
//! Step 03 (Rust) — fan-out / fan-in detection.
//!
//! Mirrors `03-fan-detection.py`. Two single-MATCH queries, no
//! self-joins. The planted `SPLITTER-1` (fan-out, 9 recipients) and
//! `MULE-1` (fan-in, 12 sources) surface deterministically.

use anyhow::Result;
use fraud_graph_traversal::load;
use std::path::PathBuf;

const FAN_THRESHOLD: i64 = 6;

fn main() -> Result<()> {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.pop();
    path.push("data");
    path.push("transfers.parquet");
    let (db, _) = load(&path)?;

    // ---- Fan-out: one source, many distinct recipients ----------------
    println!("[fan-out] Accounts sending to >= {FAN_THRESHOLD} distinct recipients:");
    let q1 = format!(
        "MATCH (a:Account)-[:SENT]->(:Transfer)-[:TO]->(b:Account) \
         RETURN a.account_id AS account, count(b) AS distinct_recipients \
         ORDER BY distinct_recipients DESC \
         LIMIT 10"
    );
    for row in db.execute(&q1)?.rows {
        let n: i64 = row.get("distinct_recipients")
            .and_then(|s| s.parse().ok()).unwrap_or(0);
        if n < FAN_THRESHOLD {
            continue;
        }
        println!("  {} → {n} recipients",
            row.get("account").cloned().unwrap_or_default());
    }

    // ---- Fan-in: one beneficiary, many distinct sources ----------------
    println!("\n[fan-in] Accounts receiving from >= {FAN_THRESHOLD} distinct sources:");
    let q2 = format!(
        "MATCH (a:Account)-[:SENT]->(:Transfer)-[:TO]->(b:Account) \
         RETURN b.account_id AS account, count(a) AS distinct_sources \
         ORDER BY distinct_sources DESC \
         LIMIT 10"
    );
    for row in db.execute(&q2)?.rows {
        let n: i64 = row.get("distinct_sources")
            .and_then(|s| s.parse().ok()).unwrap_or(0);
        if n < FAN_THRESHOLD {
            continue;
        }
        println!("  {} ← {n} sources",
            row.get("account").cloned().unwrap_or_default());
    }
    Ok(())
}
